### PR TITLE
Distributed Training For Cifar10 Example

### PR DESCRIPTION
**Pull request by Eric-Le-Ge**
This PR adds a user guide for how to use distributed training on GKE for the Cifar10 MLKit example. The workflow depends on the custom trainer executor implemented here: #2248

### CODE REVIEW

**Comment by Eric-Le-Ge** [comment]
@charlesccychen @chuanyu

File: tfx/examples/cifar10/distributed/cifar10_utils_native_keras.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
+# Lint as: python2, python3
```
Comment (charlesccychen) [inline tfx/examples/cifar10/distributed/cifar10_utils_native_keras.py:1]: Is there a way we can do this without duplicating the util module?

File: tfx/examples/cifar10/distributed/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+to upload your new image to your container registry.
+
+### Modify Pipeline and Util Files for Distributed Training
+
+First, modify the `_cifar10_root` in `cifar10_pipeline_native_keras` to point to a remote directory (For example, `gs://YOUR_GCS_PATH`),
+and upload the cifar10 base directory there when you are done modifying the files. This allows training worker pods to pull the utility files
+from your remote directory during training.
+
+Then, replace the Trainer component with the `kubernetes_trainer_executor.GenericExecutor` imported from `tfx.extensions.google_cloud_kubernetes.trainer`:
```
Comment (charlesccychen) [inline tfx/examples/cifar10/distributed/README.md:47]: Then, specify the custom Trainer executor  `kubernetes_trainer_executor.GenericExecutor` ....

File: tfx/examples/cifar10/distributed/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
+# CIFAR-10 Transfer Learning and MLKit integration Example with Distributed Training
+
+This example illustrates how to modify the base example for distributed training with a distributed node pool on Google Kubernetes Engine(GKE).
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:3]: Add space between Engine and (GKE)

File: tfx/examples/cifar10/distributed/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+If you do not have access to a GKE cluster, follow the quickstart guide [here](https://cloud.google.com/kubernetes-engine/docs/quickstart) to start one.
+
+Then, follow the instructions below for how to set up a node pool on GKE:
+https://cloud.google.com/kubernetes-engine/docs/how-to/node-pools
+
+If you wish to use GPUs in your training, you may follow this guide instead:
+https://cloud.google.com/kubernetes-engine/docs/how-to/gpus
+
+### Set Up a Custom Image for Training (Optional)
+The base cifar10 example relies on dependencies including TensorFlowJS that do not come with the default TFX installation.
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:18]: CIFAR10, here and everywhere.

File: tfx/examples/cifar10/distributed/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+Then, follow the instructions below for how to set up a node pool on GKE:
+https://cloud.google.com/kubernetes-engine/docs/how-to/node-pools
+
+If you wish to use GPUs in your training, you may follow this guide instead:
+https://cloud.google.com/kubernetes-engine/docs/how-to/gpus
+
+### Set Up a Custom Image for Training (Optional)
+The base cifar10 example relies on dependencies including TensorFlowJS that do not come with the default TFX installation. 
+If you do not wish to use a custom docker image, you can remove the TFLite model rewriting portion in `cifar10_utils_native_keras`.
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:19]: Docker (capitalised), here and everywhere

File: tfx/examples/cifar10/distributed/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+Start from a base tensorflow gpu image in Docker:
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:21]: TensorFlow, GPU

File: tfx/examples/cifar10/distributed/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+## Recommendations
+
+The following has been tested to yield the best speed-up results compared to training on a single node:
+
+- Multi-worker training with 4 or 8 replicas with 8 or 16 nodes, >= 64 GB memory
+- Single worker training with GPU (i.e. NVIDIA K80)
+
+If configured correctly, the above two configurations should yield a 60% to 70% speed up. If you need an even higher speed up,
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:84]: speed up in wall time versus single node with no GPU?

File: tfx/examples/cifar10/distributed/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+consider using multi-worker training with 1 GPU per node. However, the cost efficiency will be lower due to observed 
+bottleneck in image preprocessing.
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:86]: cost efficiency will be lower, as the training in bottlenecked by preprocessing of images on CPU (if this is indeed the case)

File: tfx/examples/cifar10/distributed/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+## Recommendations
+
+The following has been tested to yield the best speed-up results compared to training on a single node:
+
+- Multi-worker training with 4 or 8 replicas with 8 or 16 nodes, >= 64 GB memory
+- Single worker training with GPU (i.e. NVIDIA K80)
+
+If configured correctly, the above two configurations should yield a 60% to 70% speed up. If you need an even higher speed up,
+consider using multi-worker training with 1 GPU per node. However, the cost efficiency will be lower due to observed
```
Comment (chuanyu) [inline tfx/examples/cifar10/distributed/README.md:85]: How much was the speed up observed? Include the speedup here as well?

File: tfx/examples/cifar10/distributed/README.md (hunk 3 above)
Comment (charlesccychen) [reply tfx/examples/cifar10/distributed/README.md:3]: Please address @chuanyu's comments throughout.

File: tfx/examples/cifar10/distributed/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
+# CIFAR-10 Transfer Learning and MLKit integration Example with Distributed Training
+
+This example illustrates how to modify the base example for distributed training with a distributed node pool on Google Kubernetes Engine(GKE).
+
+## Instructions
+
+### Set Up a Node Pool
+The guide assumes that you have command line access to `kubectl`.
+If you do not have access to a GKE cluster, follow the quickstart guide [here](https://cloud.google.com/kubernetes-engine/docs/quickstart) to start one.
```
Comment (charlesccychen) [inline tfx/examples/cifar10/distributed/README.md:9]: Could you add the suggestion that in order to have write access to the GCS buckets in the project, you need to set up the node pool to have full service account scope access, either here or in the next paragraph?

File: tfx/examples/cifar10/distributed/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+Then, follow the instructions below for how to set up a node pool on GKE:
+https://cloud.google.com/kubernetes-engine/docs/how-to/node-pools
+
+If you wish to use GPUs in your training, you may follow this guide instead:
```
Comment (charlesccychen) [inline tfx/examples/cifar10/distributed/README.md:14]: "you may follow this guide instead" -> "follow the instructions here:"

File: tfx/examples/cifar10/distributed/README.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+Start from a base tensorflow gpu image in Docker:
+
+```
+docker run -it tensorflow/tensorflow:latest-gpu
+```
+
+Then, follow the instructions in the base example to configure your training environment. Once you are done, commit the image with:
+
+```
+docker commit [CONTAINER_ID] [new_image_name]
```
Comment (charlesccychen) [inline tfx/examples/cifar10/distributed/README.md:30]: This is a discouraged way to build containers.  Could you instead provide the contents of a Dockerfile inline?

File: tfx/examples/cifar10/distributed/cifar10_utils_native_keras.py (hunk 1 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/cifar10_utils_native_keras.py:1]: I think there are around 4 places and about 30 lines total that are different from the original util module, so it's a non-trivial change for the users to make. That being said, it should be possible to incorporate all of these changes in the README.

File: tfx/examples/cifar10/distributed/README.md (hunk 7 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/README.md:84]: Yes, this is more specific.

File: tfx/examples/cifar10/distributed/README.md (hunk 8 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/README.md:86]: Fixed.

File: tfx/examples/cifar10/distributed/README.md (hunk 9 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/README.md:85]: The most speed up observed is about 80% compared to the base single node case. I've added this data point to here as well. FYI, these are just summaries from the experimentation data shared in the internal document.

File: tfx/examples/cifar10/distributed/README.md (hunk 10 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/README.md:9]: Added as a separate paragraph.

File: tfx/examples/cifar10/distributed/README.md (hunk 12 above)
Comment (Eric-Le-Ge) [reply tfx/examples/cifar10/distributed/README.md]: Yes, I changed the instructions to build from a Dockerfile.

**Pull request closed without merging**